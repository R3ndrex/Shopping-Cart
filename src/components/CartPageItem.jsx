import { PlusIcon } from "@heroicons/react/24/solid";
import { MinusIcon } from "@heroicons/react/24/solid";
import { TrashIcon } from "@heroicons/react/24/solid";
import { useOutletContext } from "react-router-dom";

export default function CartPageItem({ item }) {
    const [setSelectedItems, _] = useOutletContext();

    function ChangeItemsAmount(item, amount) {
        setSelectedItems((prev) => {
            const existingItem = prev.find(
                (element) => element.title === item.title
            );

            if (existingItem) {
                return prev.map((element) => {
                    if (element.title === item.title) {
                        if (item.amount + amount <= 0) {
                            removeItem(item);
                        }
                        return { ...element, amount: element.amount + amount };
                    }
                    return element;
                });
            }
            return prev;
        });
    }
    const incrementAmount = (item) => ChangeItemsAmount(item, 1);

    const decrementAmount = (item) => ChangeItemsAmount(item, -1);

    function removeItem(item) {
        setSelectedItems((prev) =>
            prev.filter((element) => element.title !== item.title)
        );
    }

    return (
        <li className="item-li-shopping-cart" key={item.id}>
            <h2>{item.title}</h2>
            <img src={item.image} alt={item.title} />
            <p>{item.description}</p>
            <div className="flex gap-[1rem] text-center items-center pb-[1rem] pt-[1rem]">
                <TrashIcon
                    data-testid="trash"
                    className="h-[2rem]"
                    onClick={() => removeItem(item)}
                />
                <MinusIcon
                    data-testid="minus"
                    className="h-[2rem]"
                    onClick={() => decrementAmount(item)}
                />
                Amount: {item.amount}
                <PlusIcon
                    data-testid="plus"
                    className="h-[2rem]"
                    onClick={() => incrementAmount(item)}
                />
            </div>
        </li>
    );
}
