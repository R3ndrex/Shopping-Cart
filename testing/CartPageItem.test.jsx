import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import CartPageItem from "../src/components/CartPageItem";
import userEvent from "@testing-library/user-event";

const item = { title: "name", amount: 1, cost: 4 };
describe("CartPageItem", () => {
    it("calls increment function items", async () => {
        const incrementButton = vi.fn();
        const user = userEvent.setup();
        render(
            <CartPageItem
                item={item}
                handleDecrementAmount={vi.fn()}
                handleIncrementAmount={incrementButton}
            />
        );

        await user.click(screen.getByTestId("plus"));

        expect(incrementButton).toBeCalled();
    });

    it("calls increment function items", async () => {
        const decrementButton = vi.fn();
        const user = userEvent.setup();
        render(
            <CartPageItem
                item={item}
                handleDecrementAmount={decrementButton}
                handleIncrementAmount={vi.fn()}
            />
        );

        await user.click(screen.getByTestId("minus"));

        expect(decrementButton).toBeCalled();
    });
    it("renders correctly", () => {
        const { container } = render(
            <CartPageItem
                item={item}
                handleDecrementAmount={vi.fn()}
                handleIncrementAmount={vi.fn()}
            />
        );

        expect(container).toMatchSnapshot();
    });
});
