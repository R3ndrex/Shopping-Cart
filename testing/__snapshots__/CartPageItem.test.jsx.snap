// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartPageItem > renders correctly 1`] = `
<div>
  <li
    class="item-li-shopping-cart"
  >
    <h2>
      name
    </h2>
    <img
      alt="name"
    />
    <p />
    <div
      class="flex gap-[1rem]"
    >
      <svg
        aria-hidden="true"
        class="h-[2rem]"
        data-slot="icon"
        data-testid="minus"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.25 12a.75.75 0 0 1 .75-.75h14a.75.75 0 0 1 0 1.5H5a.75.75 0 0 1-.75-.75Z"
          fill-rule="evenodd"
        />
      </svg>
       
      Amount: 
      1
       
      <svg
        aria-hidden="true"
        class="h-[2rem]"
        data-slot="icon"
        data-testid="plus"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 3.75a.75.75 0 0 1 .75.75v6.75h6.75a.75.75 0 0 1 0 1.5h-6.75v6.75a.75.75 0 0 1-1.5 0v-6.75H4.5a.75.75 0 0 1 0-1.5h6.75V4.5a.75.75 0 0 1 .75-.75Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </li>
</div>
`;
